### PR TITLE
GH-180: Support for config driven Discovery Client

### PR DESCRIPTION
**Pull request by bijukunjummen**
#180 : Support for config driven Discovery Client.

### CODE REVIEW

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryClientAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+/**
+ * Spring Boot Auto-Configuration for Simple Properties based Discovery Client
+ *
+ * @author Biju Kunjummen
+ */
+
+@Configuration
+@ConditionalOnProperty(value = "spring.cloud.discovery.client.simple.enabled", matchIfMissing = false)
+@EnableConfigurationProperties(SimpleDiscoveryProperties.class)
+@AutoConfigureBefore(NoopDiscoveryClientAutoConfiguration.class)
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryClientAutoConfiguration.java]: I would actually want this to replace the `NoopDiscoveryClient`.

Comment (bijukunjummen) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryClientAutoConfiguration.java]: Oh, okay, safe to delete NoOpDiscoveryClient and helpers altogether or should I mark it as deprecated?

**Comment by bijukunjummen** [comment]
When you get a chance can you give it another pass @spencergibb, I have now marked NoOp* as deprecated, made SimpleDiscoveryClient as the default one.

File: spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/noop/NoopDiscoveryClientConfigurationAdditionalTests.java [hunk 2]
```diff
@@ -22,6 +23,7 @@
  */
 @RunWith(SpringRunner.class)
 @SpringBootTest(classes = NoopDiscoveryClientConfigurationAdditionalTests.App.class)
+@Ignore
```
Comment (spencergibb) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/noop/NoopDiscoveryClientConfigurationAdditionalTests.java]: Why ignored?

Comment (bijukunjummen) [inline spring-cloud-commons/src/test/java/org/springframework/cloud/client/discovery/noop/NoopDiscoveryClientConfigurationAdditionalTests.java]: These were tests related to the DiscoveryClient defaulting to the NoOp implementation, with SimpleDiscoveryClient the defaults have changed. I have now removed this test and have another test where the SimpleDiscoveryClient as default is validated.

**Comment by bijukunjummen** [comment]
another pass @spencergibb, I have refactored the tests, moved them from NoOp* to Simple* where relevant.

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryProperties.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+@ConfigurationProperties(prefix = "spring.cloud.discovery.client.simple")
+public class SimpleDiscoveryProperties {
+	private Map<String, List<SimpleServiceInstance>> instances = new HashMap<>();
+
+	public Map<String, List<SimpleServiceInstance>> getInstances() {
+		return instances;
+	}
+
+	public void setInstances(Map<String, List<SimpleServiceInstance>> instances) {
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryProperties.java]: Could this use `DefaultServiceInstance` instead? Eliminating an almost identical class?

Comment (bijukunjummen) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryProperties.java]: Yes, thought about that @spencergibb, the only reason was `DefaultServiceInstance` has all final fields and that would not work with mapping via `@ConfigurationProperties` right

Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/discovery/simple/SimpleDiscoveryProperties.java]: good point

**Comment by bijukunjummen** [comment]
Any other changes you want me to make on this PR @spencergibb

**Comment by spencergibb** [comment]
Will look again today.

**Pull request merged**